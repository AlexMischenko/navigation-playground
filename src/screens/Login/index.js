import React from 'react'
import { useDispatch } from 'react-redux'
import { useForm } from 'react-hook-form'
import { yupResolver } from '@hookform/resolvers/yup'
import { ScrollView, View, Text, Pressable, Alert } from 'react-native'

import { loginUser } from '../../services/user'
import Routes from '../../routes'
import { loginValidationSchema } from '../../utils/validators'
import theme from '../../theme'
import Layout from '../../components/Layout'
import StyledFormInput from '../../components/StyledFormInput'
import cs from './styles'

const Login = ({ navigation, route }) => {
  const dispatch = useDispatch()

  const formDefaultValues = {
    username: '',
    password: '',
  }

  const {
    control,
    handleSubmit,
    formState: { errors },
  } = useForm({
    defaultValues: formDefaultValues,
    resolver: yupResolver(loginValidationSchema),
  })

  const onSignupPress = () => {
    navigation.navigate(Routes.Signup)
  }

  const onForgotPasswordPress = () => {
    Alert.alert('Forgot password', 'Not implemeted yet :)')
  }

  const onSubmit = async (data) => {
    loginUser(data)(dispatch)
  }

  return (
    <Layout keyboardPosition>
      <ScrollView
        contentContainerStyle={theme.helpers.scrollWrap}
        keyboardShouldPersistTaps="handled"
        bounces={false}
      >
        <View style={theme.helpers.flex1} />
        <View style={theme.helpers.flex1JustifyContentCenter}>
          <Text style={[cs.titleBlock, theme.textStyles.title1]}>Login</Text>
          <View style={theme.helpers.ph20}>
            <StyledFormInput
              style={theme.helpers.mb16}
              name="username"
              control={control}
              errors={errors}
            />
            <StyledFormInput name="password" control={control} errors={errors} />
            <Pressable style={cs.submitButton} onPress={handleSubmit(onSubmit)}>
              <Text style={[theme.textStyles.title3, cs.whiteColor]}>Log In</Text>
            </Pressable>
          </View>
        </View>
        <View style={theme.helpers.alignItemsCenter}>
          <Pressable style={theme.helpers.pb20} onPress={onSignupPress}>
            <Text style={[theme.textStyles.subHeadlineBold, cs.pressableTextColor]}>Sign Up</Text>
          </Pressable>
          <Pressable style={theme.helpers.pb20} onPress={onForgotPasswordPress}>
            <Text style={theme.textStyles.subHeadline}>Forgot password ?</Text>
          </Pressable>
        </View>
      </ScrollView>
    </Layout>
  )
}

export default Login
