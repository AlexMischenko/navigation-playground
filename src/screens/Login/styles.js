import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  loginPageContainer: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: 'white',
  },
  titleBlock: {
    alignSelf: 'center',
    marginBottom: 20,
  },
  titleText: {
    fontSize: 30,
    fontWeight: 'bold',
  },
  formBlock: {
    paddingHorizontal: 30,
  },
  formInputBlock: {
    marginTop: 20,
  },
  formInput: {
    height: 60,
    width: '100%',
    paddingHorizontal: 15,
    borderRadius: 15,
    backgroundColor: '#EDF2F4',
  },
  submitButton: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 60,
    marginTop: 30,
    borderRadius: 10,
    backgroundColor: '#34C759',
  },
  submitButtonText: {
    color: 'white',
    fontSize: 14,
  },

  footerBlock: {
    alignSelf: 'center',
    alignItems: 'center',
    marginTop: 20,
  },
  footerButton: {
    paddingBottom: 20,
  },
})
