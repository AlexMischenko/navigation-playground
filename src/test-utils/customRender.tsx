import React from 'react'
import { render, RenderOptions, RenderAPI } from '@testing-library/react-native'
import { Provider } from 'react-redux'

import store from '../redux/store'

export const ProvidersWrapper: React.FC = ({ children }) => {
  return <Provider store={store}>{children}</Provider>
}

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>
): RenderAPI => render(ui, { wrapper: ProvidersWrapper, ...options })

export default customRender
