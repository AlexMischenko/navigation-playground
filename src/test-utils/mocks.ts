import type { ILoginUserResponse, ISignUpUserResponse, IGetUserResponse } from '@types'

export const mockUserResponse: ILoginUserResponse | IGetUserResponse = {
  user: { id: 1234, username: 'mock', email: 'mock@.test,com' },
  token: 'someTestToken',
}

export const mockUserRegisterResponse: ISignUpUserResponse = {
  user: { id: 1234, username: 'mock', email: 'mock@.test,com' },
}

export const mockSignUpNavigationState = {
  type: 'stack',
  index: 0,
  routeNames: ['MainStack'],
  routes: [
    {
      name: 'MainStack',
      state: {
        type: 'stack',
        index: 0,
        routeNames: ['SignUp'],
        routes: [{ name: 'SignUp' }],
      },
    },
  ],
}
