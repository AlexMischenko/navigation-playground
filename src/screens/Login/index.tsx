import React from 'react'
import { useForm, SubmitHandler } from 'react-hook-form'
import { yupResolver } from '@hookform/resolvers/yup'
import { ScrollView, View, Text, Pressable, Alert } from 'react-native'

import type { ILoginScreenProps, ILoginForm } from '@types'
import { useDispatch, useTheme, useStyleSheet } from '@hooks'
import { loginUser } from '@services/user'
import { loginValidationSchema } from '@utils/validators'
import Routes from '@routes'
import Layout from '@components/Layout'
import StyledFormInput from '@components/StyledFormInput'
import styles from './styles'

const Login: React.FC<ILoginScreenProps> = ({ navigation }) => {
  const theme = useTheme()
  const cs = useStyleSheet(styles)
  const dispatch = useDispatch()

  const formDefaultValues = {
    username: '',
    password: '',
  }

  const {
    control,
    handleSubmit,
    formState: { errors },
  } = useForm<ILoginForm>({
    defaultValues: formDefaultValues,
    resolver: yupResolver(loginValidationSchema),
  })

  const onSignupPress = () => {
    navigation.navigate(Routes.SignUp)
  }

  const onForgotPasswordPress = () => {
    dispatch({ type: 'HELLO_SAGA', payload: 5 })
    // Alert.alert('Forgot password', 'Not implemeted yet :)')
  }

  const onSubmit: SubmitHandler<ILoginForm> = (data) => {
    // loginUser(data)(dispatch)

    dispatch({ type: 'LOGIN_REQUEST', payload: data })
  }

  return (
    <Layout keyboardPosition>
      <ScrollView
        contentContainerStyle={theme.helpers.scrollWrap}
        keyboardShouldPersistTaps="handled"
        bounces={false}
      >
        <View style={theme.helpers.flex1} />
        <View style={theme.helpers.flex1JustifyContentCenter}>
          <Text style={[cs.titleBlock, theme.textStyles.title1]}>Login</Text>
          <View style={theme.helpers.ph20}>
            <StyledFormInput
              style={theme.helpers.mb16}
              name="username"
              control={control}
              errors={errors}
            />
            <StyledFormInput name="password" control={control} errors={errors} />
            <Pressable style={cs.submitButton} onPress={handleSubmit(onSubmit)}>
              <Text style={[theme.textStyles.title3, cs.whiteColor]}>Log In</Text>
            </Pressable>
          </View>
        </View>
        <View style={theme.helpers.alignItemsCenter}>
          <Pressable style={theme.helpers.pb20} onPress={onSignupPress}>
            <Text style={[theme.textStyles.subHeadlineBold, cs.pressableTextColor]}>Sign Up</Text>
          </Pressable>
          <Pressable style={theme.helpers.pb20} onPress={onForgotPasswordPress}>
            <Text style={theme.textStyles.subHeadline}>Forgot password ?</Text>
          </Pressable>
        </View>
      </ScrollView>
    </Layout>
  )
}

export default Login
