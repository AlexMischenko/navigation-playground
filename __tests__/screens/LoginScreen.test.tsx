import * as React from 'react'
import { fireEvent, act } from '@testing-library/react-native'

import Login from '@screens/Login'
import { mockUserResponse } from '../../src/test-utils/mocks'
import RootNavigator from '../../src/navigator'
import customRender from '../../src/test-utils/customRender'

const nav: any = {}

jest.mock('@api/user', () => ({
  loginUser: jest.fn(() => Promise.resolve(mockUserResponse)),
  getUser: jest.fn(() => Promise.resolve(mockUserResponse)),
}))

describe('Testing @screens/Login', () => {
  it('`Login` screen renders correctly', () => {
    const wrapper = customRender(<Login navigation={nav} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('`Login` screen has "username" input', () => {
    const { getByText } = customRender(<Login navigation={nav} />)
    const usernameInputTitle = getByText('username')

    expect(usernameInputTitle).toBeTruthy()
  })

  it('`Login` screen has "password" input', () => {
    const { getByText } = customRender(<Login navigation={nav} />)
    const passwordInputTitle = getByText('password')

    expect(passwordInputTitle).toBeTruthy()
  })

  it('`Login` screen has submit button', () => {
    const { getByText } = customRender(<Login navigation={nav} />)
    const submitButton = getByText('Log In')

    expect(submitButton).toBeTruthy()
  })

  it('`Login` screen navigates to `SignUp` screen correctly', async () => {
    const { getByText } = customRender(<RootNavigator />)
    const goToSignUpButton = getByText('Sign Up')

    fireEvent.press(goToSignUpButton)

    const registerText = getByText('Register')

    expect(registerText).toBeTruthy()
  })

  it('`Login` screen navigates to `Home` screen after form submit correctly', async () => {
    const { getByText, getByA11yLabel } = customRender(<RootNavigator />)
    const usernameInput = getByA11yLabel('username input')
    const passwordInput = getByA11yLabel('password input')
    const submitButton = getByText('Log In')

    fireEvent.changeText(usernameInput, 'test')
    fireEvent.changeText(passwordInput, 'test')
    await act(async () => {
      fireEvent.press(submitButton)
    })

    const homepageText = getByText('Page1')

    expect(homepageText).toBeTruthy()
  })
})
