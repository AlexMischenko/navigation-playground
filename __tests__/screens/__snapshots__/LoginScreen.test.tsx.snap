// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing @screens/Login \`Login\` screen renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "top",
      "bottom",
    ]
  }
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "white",
        "paddingHorizontal": 0,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "overflow": "hidden",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "bottom": 0,
          },
        ]
      }
    >
      <RCTScrollView
        bounces={false}
        contentContainerStyle={
          Object {
            "flexGrow": 1,
            "justifyContent": "space-between",
            "paddingHorizontal": 36,
          }
        }
        keyboardShouldPersistTaps="handled"
      >
        <View>
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          />
          <View
            style={
              Object {
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "alignSelf": "center",
                    "marginBottom": 36,
                  },
                  Object {
                    "color": "#2B2D42",
                    "fontSize": 50.72463768115942,
                    "letterSpacing": 0.6702898550724637,
                    "lineHeight": 61.594202898550726,
                  },
                ]
              }
            >
              Login
            </Text>
            <View
              style={
                Object {
                  "paddingHorizontal": 36,
                }
              }
            >
              <View
                style={
                  Object {
                    "marginBottom": 29,
                  }
                }
              >
                <View>
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#2B2D42",
                          "fontSize": 28.985507246376812,
                          "letterSpacing": -0.5797101449275363,
                          "lineHeight": 38.04347826086956,
                          "textTransform": "capitalize",
                        },
                        Object {
                          "marginBottom": 14.5,
                        },
                        undefined,
                      ]
                    }
                  >
                    username
                  </Text>
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "#ffffff",
                          "borderRadius": 25.5,
                          "elevation": 3,
                          "flexDirection": "row",
                          "height": 116,
                          "justifyContent": "space-between",
                          "paddingLeft": 31,
                          "paddingRight": 30,
                          "shadowColor": "#8D99AE",
                          "shadowOffset": Object {
                            "height": 0,
                            "width": 0,
                          },
                          "shadowOpacity": 0.15,
                          "shadowRadius": 18,
                          "width": "100%",
                        },
                        undefined,
                      ]
                    }
                  >
                    <TextInput
                      accessibilityLabel="username input"
                      allowFontScaling={true}
                      autoCapitalize="none"
                      onBlur={[Function]}
                      onChangeText={[Function]}
                      rejectResponderTermination={true}
                      style={
                        Array [
                          Object {
                            "flex": 1,
                          },
                          undefined,
                        ]
                      }
                      underlineColorAndroid="transparent"
                      value=""
                    />
                  </View>
                </View>
              </View>
              <View>
                <View>
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#2B2D42",
                          "fontSize": 28.985507246376812,
                          "letterSpacing": -0.5797101449275363,
                          "lineHeight": 38.04347826086956,
                          "textTransform": "capitalize",
                        },
                        Object {
                          "marginBottom": 14.5,
                        },
                        undefined,
                      ]
                    }
                  >
                    password
                  </Text>
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "#ffffff",
                          "borderRadius": 25.5,
                          "elevation": 3,
                          "flexDirection": "row",
                          "height": 116,
                          "justifyContent": "space-between",
                          "paddingLeft": 31,
                          "paddingRight": 30,
                          "shadowColor": "#8D99AE",
                          "shadowOffset": Object {
                            "height": 0,
                            "width": 0,
                          },
                          "shadowOpacity": 0.15,
                          "shadowRadius": 18,
                          "width": "100%",
                        },
                        undefined,
                      ]
                    }
                  >
                    <TextInput
                      accessibilityLabel="password input"
                      allowFontScaling={true}
                      autoCapitalize="none"
                      onBlur={[Function]}
                      onChangeText={[Function]}
                      rejectResponderTermination={true}
                      style={
                        Array [
                          Object {
                            "flex": 1,
                          },
                          undefined,
                        ]
                      }
                      underlineColorAndroid="transparent"
                      value=""
                    />
                  </View>
                </View>
              </View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#34C759",
                    "borderRadius": 25.5,
                    "height": 108.5,
                    "justifyContent": "center",
                    "marginTop": 54.5,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2B2D42",
                        "fontSize": 36.231884057971016,
                        "letterSpacing": 0.6884057971014492,
                        "lineHeight": 45.289855072463766,
                      },
                      Object {
                        "color": "#ffffff",
                      },
                    ]
                  }
                >
                  Log In
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
              }
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "paddingBottom": 36,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#2B2D42",
                      "fontSize": 27.17391304347826,
                      "fontWeight": "600",
                      "letterSpacing": -0.43478260869565216,
                      "lineHeight": 36.231884057971016,
                    },
                    Object {
                      "color": "#4A80F0",
                    },
                  ]
                }
              >
                Sign Up
              </Text>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "paddingBottom": 36,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#2B2D42",
                    "fontSize": 27.17391304347826,
                    "letterSpacing": -0.43478260869565216,
                    "lineHeight": 36.231884057971016,
                  }
                }
              >
                Forgot password ?
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</RNCSafeAreaView>
`;
